### PR TITLE
[Enhancement] mv rewrite predicate split support multi range

### PR DESCRIPTION
**Pull request by liuyufei9527**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 135 / 140 (96.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|100|104|96.15%|[129, 213, 214, 234]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java|33|34|97.06%|[82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java [hunk 1]
```diff
@@ -75,33 +88,196 @@ public static PredicateSplit splitPredicate(ScalarOperator predicate) {
...
+                if (predicate.isOr()) {
+                    // (a = 1 and b = 1) or (a = 2 and b = 2) can't convert to range
+                    if (childRange.size() > 1) {
+                        return null;
+                    }
+                    if (columnRangeSet.isEmpty()) {
+                        columnRangeSet.putAll(childRange);
+                    } else {
+                        for (ColumnRefOperator column : childRange.keySet()) {
+                            if (!columnRangeSet.containsKey(column)) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:196]: So here we can only support one column's or predicates:
MV: a = 1 or a = 2 or a = 3
Query: a = 1 or a = 2

Query: From MV where a = 1 or a = 2


Just wonder can we support this?

MV : a = 1 or b = 1 or c = 1
Query: a = 1 or b = 1

=>
Query : From MV where a = 1 or b = 1?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java [hunk 2]
```diff
@@ -75,33 +88,196 @@ public static PredicateSplit splitPredicate(ScalarOperator predicate) {
...
-        List<ScalarOperator> predicateConjuncts = Utils.extractConjuncts(predicate);
-        List<ScalarOperator> columnEqualityPredicates = Lists.newArrayList();
         List<ScalarOperator> rangePredicates = Lists.newArrayList();
+        RangeExtractor extractor = new RangeExtractor();
+        Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>> rangeSet =
+                predicate.accept(extractor, new RangeExtractorContext());
+        if (rangeSet != null) {
+            rangeSet.forEach((column, range) -> rangePredicates.add(rangeSetToExpr(range, column)));
+        } else if (extractor.columnEqualityPredicates.isEmpty() && extractor.residualPredicates.isEmpty()) {
+            extractor.residualPredicates.add(predicate);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:98]: In this case `extractor.residualPredicates` must be empty too?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java [hunk 3]
```diff
@@ -75,33 +88,196 @@ public static PredicateSplit splitPredicate(ScalarOperator predicate) {
...
+            return null;
+        }
+
+        @Override
+        public Map<ColumnRefOperator, TreeRangeSet<ConstantOperator>> visitBinaryPredicate(
+                BinaryPredicateOperator predicate, RangeExtractorContext context) {
+            ScalarOperator left = predicate.getChild(0);
+            ScalarOperator right = predicate.getChild(1);
+            if (left.isColumnRef() && right.isConstantRef()) {
+                if (predicate.getBinaryType().isNotEqual()) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:144]: still cannot change to `!=` to range predicates?

File: fe/fe-core/src/test/java/com/starrocks/planner/MaterializedViewTest.java [hunk 4]
```diff
@@ -2655,7 +2655,7 @@ public void testRangePredicates2() {
             testRewriteOK(mv, "select deptno as col1, empid as col2, emps.locationid as col3 from emps " +
                     " left join locations on emps.locationid = locations.locationid where emps.deptno < 10 or emps.deptno > 10");
             // TODO: support OrPredicates as Range Predicates.
-            testRewriteFail(mv, "select deptno as col1, empid as col2, emps.locationid as col3 from emps " +
+            testRewriteOK(mv, "select deptno as col1, empid as col2, emps.locationid as col3 from emps " +
                     " left join locations on emps.locationid = locations.locationid where emps.deptno < 5 or emps.deptno > 20");
         }
```
Comment (LiShuMing) [inline fe/fe-core/src/test/java/com/starrocks/planner/MaterializedViewTest.java:2660]: Greate job! can you add more complex tests here to display we can support `or`  predicates?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java (hunk 1 above)
Comment (liuyufei9527) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:196]: In residual subsumption test allreadly support this, I will add some test case.

Comment (liuyufei9527) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:196]: mv: a > 0 OR b > 0
query: a > 5
this case not support

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java (hunk 3 above)
Comment (liuyufei9527) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java]: already supported

File: fe/fe-core/src/test/java/com/starrocks/planner/MaterializedViewTest.java (hunk 4 above)
Comment (liuyufei9527) [reply fe/fe-core/src/test/java/com/starrocks/planner/MaterializedViewTest.java:2657]: add some case in MaterializedViewTest#testRangePredicates3.
you can reply here if want to add some case.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 138 / 142 (97.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|103|106|97.17%|[197, 198, 228]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java|33|34|97.06%|[82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 138 / 142 (97.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java|103|106|97.17%|[197, 198, 228]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java|33|34|97.06%|[82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|

**Review by ABingHuang — CHANGES_REQUESTED** [review]

**Review by ABingHuang — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.5 branch-3.0 branch-3.1

**Pull request merged**